accept .ts route files if supported

mielp opened the pull request:
See #14 

- [x] All tests pass
- [ ] Changes have been documented (TODO)

I would write new tests, but I am not sure how to do so, since there's a `tsc` step when running tests today.

A _workaround_ would be to have a command that runs the existing tests with `ts-node` enabled. This way, all current tests would be ran once in JS, and once in TS.

According to the [AVA documentation for TypeScript](https://github.com/avajs/ava/blob/main/docs/recipes/typescript.md), adding something like that to the `package.json` would work:

```json5
{
  "scripts": {
    "test": "ava test-build/**/*.spec.js", // no change here
    "test:ts": "ava test/**/*.spec.ts" // new script!
  },
  "devDependencies": {
    "ts-node": "^10.2.0", // new dependency
  },
  "ava": {
    "extensions": [ // tell ava to accept both js (test) and ts (test:ts)
      "js", "ts"
    ],
    "require": [ // use ts-node/register in order to transpile TS on the fly
      "ts-node/register"
    ]
  }
}
```

But then we would need to run both `npm run test` and `npm run test:ts`.

mielp commented:
What do you think @yonathan06? 🙂 If needed I can document the change and bump the version number in this very same PR as well (v2.4.0? new feature, no breaking change as far as I can tell, so a major update seems unnecessary)

Bug introduced by this PR: if a directory contains both `.js` and `.d.ts` files (as is often the case with compiled projects), and `ts-node` support is detected, both files will be loaded, and loading a `.d.ts` file might break things. Declaration files should be ignored (TODO).

yonathan06 commented:
I think it is a good direction :)
Please write tests cases to cover it. For example a server that uses `ts-node` and has `.d.ts` files in the routes folder.

mielp commented:
> 
> 
> I think it is a good direction :)
> Please write tests cases to cover it. For example a server that uses `ts-node` and has `.d.ts` files in the routes folder.

Sure! I have one question though: what do you think would be the best approach to write those tests?

Testing with `.d.ts` files in the output folder does not require additional tests: we can tellTypeScript to emit `.d.ts` files for us. After adding [`"declaration": true`](https://www.typescriptlang.org/tsconfig/#declaration) to the `compilerOptions` of `tsconfig.test.json`:

- Without TypeScript support: 23 tests OK ✔️
- With TypeScript enabled:  
  `'fastify-now: error registering routers: Cannot use import statement outside a module'` ❌
- With TypeScript enabled, after fix to ignore `.d.ts` files: 23 tests OK ✔️

However to test that requiring `.ts` files directly work as expected, we need to run tests on the uncompiled `test` directory, and not on the `test-build` directory. I think adding a new test script to the `package.json` file, as described above (in the PR description), is the best approach. Would that be okay @yonathan06?

yonathan06 commented:
Looking great! Now you can add a test case in the `test` directory (something like `typescript-server`) and add the typescript test command to `.github/workflows/ci.yaml`

mielp commented:
Sure, I added the script there!

I am not sure what a `typescript-server` would do, since `regular-server` and others are already tested in TypeScript mode by the new command, and that seems to cover all existing behavior:

- `.js` files are tested by the regular `npm test` command on all 3 servers (`regular-server`, `error-server`, `regular-server-windows` are all tested after compilation)
- `.d.ts`  files are also tested by the regular `npm test` command
- `.ts`  files are already tested by the new command `npm run test:ts` on all 3 servers as well (`regular-server`, `error-server`, `regular-server-windows` are all tested uncompiled, directly in TypeScript, with `ts-node`)
 
I added a new test server though (a copy of `regular-server`) with two of the route files renamed to `.js`, so that it can also test a server with "mixed" routes (there are both TS files and JS files). For normal (compiled) tests to work, this requires setting `allowJs` to `true` in the `tsconfig.json` file.

yonathan06 commented:
Aren't you checking also to see that when running with `ts-node` it ignores the `.d.ts` files? if so, I suggest to add some `.d.ts` files to the test server and check that the routes there are not registered

mielp commented:
Done! There are `.d.ts` in the source directory for `regular-server-ts-and-js`.

When running a buggy implementation that attempts to import `.d.ts` files, the before hook (server initialization) will fail because declaration files cannot be imported at all and `ts-node` throws:

```
error registering routers: Unable to require file: test/regular-server-ts-and-js/test-server/routes/book.d.ts
```

With the current implementation (ignore `.d.ts`), all tests pass.

yonathan06 commented:
Great work @mielp !
